Add CMS page view and duplicate functionality

Hlavtox opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Read below
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #28079, Fixes #29674
| Related PRs       | -
| How to test?      | -
| Possible impacts? | No
| Sponsor company   | [TRENDO s.r.o.](https://www.trendo.cz)

### Description
- Adds functionality to view CMS page in front office from admin CMS edit screen.
  - It adds a view icon to CMS page list.
  - It adds a view icon to CMS edit page header.
- Adds functionality to duplicate a CMS page. 
  - Click hamburger menu next to CMS page, click duplicate, confirm.
  - A copy of the page with a new title `copy of pagename` will get saved and you will get redirected to edit screen of this CMS page
- It introduces a CmsRepository where this duplication is handled. 
- It introduces EntityNameDuplicator which handles prepending the "Copy " word before product and CMS names.

### READ THIS QA
- The misaligned icon in CMS page list is not in a scope of this PR, will be solved separately. Issue https://github.com/PrestaShop/PrestaShop/issues/29692

**How does it look**
![192231866-2fada57d-0aa5-45de-af0e-470c077a180e](https://user-images.githubusercontent.com/6097524/197173194-f4fd1c2f-1eb3-4500-baf0-ffaed7130bfd.jpg)
![192231871-16bec2de-eac8-4571-8ea9-4d4fbd22ef73](https://user-images.githubusercontent.com/6097524/197173196-934e592c-8dfa-41a2-9b2e-be482e40468f.jpg)
![191951574-9a568e33-2881-4493-8434-d474d842ece2](https://user-images.githubusercontent.com/6097524/197173319-db51bb12-bc48-4e82-b240-8d047947b8ae.png)

Hlavtox committed:
Add CMS page view and duplicate functionality

kpodemski commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -310,6 +310,20 @@ public function editAction(Request $request, $cmsPageId)
+     *
+     * @return Response
+     */
+    public function viewAction(Request $request, $cmsPageId)
```
```suggestion
    public function viewAction(Request $request, int $cmsPageId)
```

kpodemski commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -310,6 +310,20 @@ public function editAction(Request $request, $cmsPageId)
 
+    /**
+     * Preview a CMS page in front office
+     *
```
```suggestion
     *
     * @param Request $request
```

kpodemski commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -310,6 +310,20 @@ public function editAction(Request $request, $cmsPageId)
+    public function viewAction(Request $request, $cmsPageId)
+    {
+        /** @var EditableCmsPage $editableCmsPage */
+        $editableCmsPage = $this->getQueryBus()->handle(new GetCmsPageForEditing((int) $cmsPageId));
```
```suggestion
        $editableCmsPage = $this->getQueryBus()->handle(new GetCmsPageForEditing($cmsPageId));
```

kpodemski commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -310,6 +310,20 @@ public function editAction(Request $request, $cmsPageId)
+    {
+        /** @var EditableCmsPage $editableCmsPage */
+        $editableCmsPage = $this->getQueryBus()->handle(new GetCmsPageForEditing((int) $cmsPageId));
+        return $this->redirect($editableCmsPage->getPreviewUrl());
```
```suggestion

        return $this->redirect($editableCmsPage->getPreviewUrl());
```

Hlavtox commented:
@kpodemski Ready for review

ElodieOS commented:
Hi @Hlavtox ! I don't think the position of the preview button is correct here. 

Before preview, you need to save so putting it near the save button is better from an ux point of view. For UI, we reuse what already exists in the product page to keep consistency

The behavior: when you click on the preview button, a new tab opens and each time you press this button, the tab updates (rather than opening yet another new tab).
Of course, you have to save the changes first to be able to preview the changes

Possible to modify, as below?
![image](https://user-images.githubusercontent.com/113457393/192006177-3b95138d-9e6a-4088-9dea-a167e5c5bc1d.png)

Hlavtox commented:
Hi, I changed it a bit by your suggestions:
- Preview now opens in a new tab. (_a new tab opens and each time you press this button, the tab updates (rather than opening yet another new tab_ - not possible, product page always opens a new tab).
- Modified the controller so it stays on the same page when saving the CMS.
- Moved the button to bottom, adapted to both desktop and mobile view.
- Changed the wording from Preview to View. Why? There is no **preview** functionality (as is on product page), we can only **view** the finished, published page the same way as customers do.
- I did not remove the Save and Preview button, we cannot make any BC breaks for next major version.

**Desktop:**
![desktop](https://user-images.githubusercontent.com/6097524/192231866-2fada57d-0aa5-45de-af0e-470c077a180e.jpg)

**Mobile:**
![view](https://user-images.githubusercontent.com/6097524/192231871-16bec2de-eac8-4571-8ea9-4d4fbd22ef73.jpg)

Julievrz commented on src/Core/Grid/Definition/Factory/CmsPageDefinitionFactory.php:
```diff
@@ -176,6 +188,21 @@ protected function getColumns()
+                        'route_param_name' => 'cmsPageId',
+                        'route_param_field' => 'id_cms',
+                        'modal_options' => new ModalOptions([
+                            'title' => $this->trans('Duplicate page', [], 'Admin.Actions'),
```
This wording is new. Do we really need to create a new string? Isn't "duplicate" enough? 

@l-delin

FabienPapet commented on src/Adapter/CMS/Page/CommandHandler/DuplicateCmsPageHandler.php:
```diff
@@ -0,0 +1,109 @@
+
+namespace PrestaShop\PrestaShop\Adapter\CMS\Page\CommandHandler;
+
+use CMS;
```
Wondering if we should use the adapter entity namespace (for all entities)?

Hlavtox replied:
Hi, I don't know, I just folled the original structure, you can discuss this with @eternoendless. 👍

eternoendless replied:
We don't use the Adapter namespace anymore, any new class goes in Core or PrestaShopBundle depending on its nature.

zuk3975 commented on src/Core/Domain/CmsPage/CommandHandler/DuplicateCmsPageHandlerInterface.php:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * @param DuplicateCmsPageCommand $command
+     */
+    public function handle(DuplicateCmsPageCommand $command);
```
Missing return type of CmsPageId (and in docblock too)

Hlavtox replied:
Done ✅

zuk3975 commented on src/Adapter/CMS/Page/CommandHandler/DuplicateCmsPageHandler.php:
```diff
@@ -0,0 +1,109 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(DuplicateCmsPageCommand $command)
```
```suggestion
    public function handle(DuplicateCmsPageCommand $command): CmsPageId
```

Hlavtox replied:
Done ✅

zuk3975 commented on src/Adapter/CMS/Page/CommandHandler/DuplicateCmsPageHandler.php:
```diff
@@ -0,0 +1,109 @@
+        $source_cms = new CMS((int) $source_cms_id);
+
+        // Prepare and save new page
+        $new_cms = new CMS();
+        $new_cms->id_cms_category = $source_cms->id_cms_category;
+        $new_cms->indexation = $source_cms->indexation;
+        $new_cms->active = false;
+        $new_cms->meta_title = $this->getNewPageTitles($source_cms->meta_title);
+        $new_cms->content = $source_cms->content;
+        $new_cms->head_seo_title = $source_cms->head_seo_title;
+        $new_cms->meta_description = $source_cms->meta_description;
+        $new_cms->meta_keywords = $source_cms->meta_keywords;
+        $new_cms->link_rewrite = $source_cms->link_rewrite;
+        $new_cms->add();
```
did you consider setting the sourceCms->id to null instead of all this manual duplication (I think it should work the same)? maybe you would only have to hardcode the `active = false`

Hlavtox replied:
Done ✅

zuk3975 commented on src/Adapter/CMS/Page/CommandHandler/DuplicateCmsPageHandler.php:
```diff
@@ -0,0 +1,109 @@
+    {
+        // Get current data
+        $source_cms_id = $command->getCmsPageId()->getValue();
+        $source_cms = new CMS((int) $source_cms_id);
```
use $camelCase variables

Hlavtox replied:
Done ✅

zuk3975 commented on src/Adapter/CMS/Page/CommandHandler/DuplicateCmsPageHandler.php:
```diff
@@ -0,0 +1,109 @@
+     */
+    public function handle(DuplicateCmsPageCommand $command)
+    {
+        // Get current data
+        $source_cms_id = $command->getCmsPageId()->getValue();
+        $source_cms = new CMS((int) $source_cms_id);
+
+        // Prepare and save new page
+        $new_cms = new CMS();
+        $new_cms->id_cms_category = $source_cms->id_cms_category;
+        $new_cms->indexation = $source_cms->indexation;
+        $new_cms->active = false;
+        $new_cms->meta_title = $this->getNewPageTitles($source_cms->meta_title);
+        $new_cms->content = $source_cms->content;
+        $new_cms->head_seo_title = $source_cms->head_seo_title;
+        $new_cms->meta_description = $source_cms->meta_description;
+        $new_cms->meta_keywords = $source_cms->meta_keywords;
+        $new_cms->link_rewrite = $source_cms->link_rewrite;
+        $new_cms->add();
```
so it would be like 
```suggestion
        // Get current data
        $sourceCmsId = $command->getCmsPageId()->getValue();
        $sourceCms = new CMS((int) $sourceCmsId);

        // reset the id so the new cms is saved having all the other properties copied from sourceCms
        $sourceCms->id = null;
        $sourceCms->active = false;
        $sourceCms->add();
```

zuk3975 replied:
oh and by the way, we have this convention of using repositories for object models since product page.
So you should probably introduce CmsPageRepository which extends AbstractObjectModelRepository and use provided reusable methods to get your CMS object model (provided reusable methods loads it with wrapped try-catch, so it will throw exception if such cms doesn't exist).
Then you can implement dedicated method (and move this duplication logic) inside that repository and your handler would look smth like:
```
$newCms = $this->cmsPageRepository->duplicate($command->getCmsPageId);

return new CmsPageId((int) $newCms->id);
```

You can see repository examples by searching for implementations of AbstractObjectModelRepository::getObjectModel() method

Hlavtox replied:
@zuk3975 Didn't go for this one, the code is now very short as I used the way of resetting ID as you proposed. :-) It's a good idea, it also duplicates some custom CMS page properties that modules could add. 👍 CLEVER BOY

zuk3975 replied:
Yes, the code is short and looks good, but its not only about that. The AbstractObjectModelRepository offers some methods like addObjectModel and getObjectModel that are wrapped in a try catch, so it ensures the class is loaded correctly and throws related Core exception if something goes wrong (whereas now, it will throw legacy PrestaShop exception which is not handled in the controller). Btw it keeps things similar as in other recently migrated pages like the Product one. I would really love to see it implemented here :smile: 
If don't have time for it or you don't want to implement it for some reason by your own, I can do it by contributing to your branch, no problem, just let me know :v:

zuk3975 commented on src/Core/Domain/CmsPage/CmsPageSettings.php:
```diff
@@ -0,0 +1,48 @@
+    /**
+     * Bellow constants define maximum allowed length of cms page properties
+     */
+    public const MAX_TITLE_LENGTH = 255;
```
:+1:

zuk3975 commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -310,6 +314,54 @@ public function editAction(Request $request, $cmsPageId)
+     *
+     * @return Response
+     */
+    public function viewAction(Request $request, int $cmsPageId)
```
```suggestion
    public function viewAction(Request $request, int $cmsPageId): RedirectResponse
```

Hlavtox replied:
Done ✅

zuk3975 commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -310,6 +314,54 @@ public function editAction(Request $request, $cmsPageId)
+     * @param Request $request
+     * @param int $cmsPageId
+     *
+     * @return Response
```
```suggestion
     * @return RedirectResponse
```

Hlavtox replied:
Done ✅

zuk3975 reviewed (changes requested):
some return types missing and the concept of CmsPageRepository inside the handler (left comments, hopefully its clear).
Otherwise good job :muscle: :+1:

Hlavtox commented:
Ready for review! :-) @zuk3975 @kpodemski

zuk3975 commented on src/PrestaShopBundle/Entity/Repository/CmsPageRepository.php:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return array<int, string>
+     */
+    private function getNewPageTitles(array $oldCmsPageLocalizedTitles): array
```
good job introducing the repo :muscle: 
Though about this method - it is same as in product page duplication. So as you mentioned in private, maybe we could think of a way to reuse it in both places?

Could we have some AbstractDuplicator service that could then be reused by both Product and Cms to rename these localized titles?
I need help deciding this @jolelievre :pray:

jolelievre replied:
I guess it would even be worth having a `LocalizedFieldDuplicator` service instead that is injected in both those services and can be used via a method:

```php
public function duplicateLocalizedValues(array $localizedValues): array
```

And this service is the one in which we inject the translator. This way it can also be unit tested and resued in many other use cases potentially.

Hlavtox replied:
Done ✅

zuk3975 commented on src/PrestaShopBundle/Entity/Repository/CmsPageRepository.php:
```diff
@@ -0,0 +1,110 @@
+     * @throws CmsPageNotFoundException
+     * @throws CannotDuplicateCmsPageException
+     */
+    public function duplicate(CmsPageId $cmsPageId): CmsPageId
```
I think we will need to do it as in ProductRepository. So this method should accept whole CmsPage object model and only handle the unsetting of id and saving the obj model, while leaving the rest of the logic for dedicated CmsDuplicator service (yeah, more code, but also more SOLID).
Because the "getNewPageTitles" logic does not belong to the repository pattern, so it should stay outside.

Hlavtox replied:
Done ✅

zuk3975 commented on src/Core/Util/Entity/EntityNameDuplicator.php:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return string
+     */
+    public function getNewName(string $currentName, int $maxLength, int $idLang): string
```
```suggestion
    public function getNewName(string $currentName, int $maxLength, int $langId): string
```

zuk3975 commented on src/Core/Util/Entity/EntityNameDuplicator.php:
```diff
@@ -0,0 +1,100 @@
+    public function getNewName(string $currentName, int $maxLength, int $idLang): string
+    {
+        // Add copy word to the string
+        $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($idLang));
```
following above suggestion
```suggestion
        $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```

zuk3975 commented on src/Core/Util/Entity/EntityNameDuplicator.php:
```diff
@@ -0,0 +1,100 @@
+use PrestaShop\PrestaShop\Core\Util\String\StringModifierInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class EntityNameDuplicator
```
Could you add some simple unit test for this class?

jolelievre replied:
Agreed, you can find some example in our code base that mock the translator (basically mock the `trans` method and return the first argument - the wording)

zuk3975 commented on src/PrestaShopBundle/Entity/Repository/CmsPageRepository.php:
```diff
@@ -0,0 +1,84 @@
+
+        // Reset object ID, make it inactive and save it
+        $cms->id = null;
+        $cms->meta_title = $this->entityNameDuplicator->getNewLocalizedNames(
```
entity name duplicator is a nice service, but what I am still in doubt (as we discussed privately), that this duplicator service doesn't really fit in the repository pattern.
If we keep following product example, there should be a CmsPageDuplicator service responsible for changing the properties and then repository basically only saves it (thats what its for - persistence).

Though I feel Im going too far with my requests and I even think that I made it wrong in product repository too. Instead of having a redundant "duplicate" method in repository, product Id must have been unset in the duplicator and then common ProductRepository::add() method should have been called.
I need help @jolelievre ! could you help me decide what we do with this PR, because it really looks good and clean, just these little details left (which are kind looping around cuz of my mistakes)

zuk3975 replied:
btw the validation is missing :sweat_smile:

jolelievre replied:
I think it's ok to have the duplicate method here, for product it's a bit different because the domain is more complex it's not only about duplicating an entity it's also all its relations, and the repository is already quite big, so it makes sense to divide it into several services

zuk3975 replied:
ok, glad to hear. Though we are missing the validation for this cms creation. Not sure if its a big deal, just if some error ever happens, the legacy PrestashopException will bubble out to the surface :zany_face: 

Im not requesting this, but just to imagine what I thought would have been ideal (before I forget :smile:):
```
class CmsPageRepository
{
    ...
    // this coud have been reused in addCmsPageHandler
    public function add(CmsPage $cmsPage): CmsPage
    {
        // as in other recent pages, validates properties one by one and ensures no legacy exception bubbles out
        $this->cmsPageValidator->validate($cmsPage);
        $this->addObjectModel($cmsPage...);
        
        return $cmsPage;
    }
 }   
    
class CmsPageDuplicator
{
    ...
    public function duplicate(CmsPageId $cmsPageId): CmsPage
    {
        $cmsPage = $this->cmsPageRepository->get($cmsPageId);
        $cmsPage->id = null;
        $cmsPage->name = $this->nameDuplicator->getNewLocalizedName(...);
        $cmsPage->{some other changes if needed)...;
        
        return $this->cmsPageRepository->add($cmsPage);
    }
}
```
So what we would get from that:
1. all props are well validated
2. CmsPageRepository has a reusable "add" method
3. Responsibilities are clear - duplicator duplicates, repository persists (so we don't have additional logic inject into repository).
Negative stuff:
1. a little more work to do (especially knowing that all those repository services ain't designed for long term, as they are still relying on object model)

jolelievre commented on src/Adapter/Product/ProductDuplicator.php:
```diff
@@ -209,6 +221,9 @@ private function duplicateProduct(Product $product): Product
     }
 
     /**
+     * @deprecated since 8.1 and will be removed in next major version.
```
The method is private so you don't need to deprecate it you can remove it And also remove the `stringModifier` dependency I believe since it won't be used anymore

Hlavtox replied:
Done ✅

jolelievre commented on src/Core/Util/Entity/EntityNameDuplicator.php:
```diff
@@ -0,0 +1,100 @@
+            $langId = (int) $langId;
+
+            // Add copy word to the string
+            $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```
For now the hard-coded message and namespace is fine since it's only used in the Catalog domain, but maybe we'll have to make it configurable in the future For now it's enough

zuk3975 replied:
also as noticed the static Language::getLocaleById method is hard to test.
Instead we may want to inject related repository, but I see we don't have one implemented yet.
I was thinking one option may be to add 
1. `public function get(LanguageId $languageId): LanguageInterface` in
https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Language/LanguageRepositoryInterface.php ? But that would require loading whole Language object model under the hood.
2. other choice is to create a LanguageRepository as all other repos we have and wrap that localeById method inside normal public function.
But these would again increase the scope, so If possilbe to test the static method, I think its ok for now.
Wdyt @jolelievre?

Hlavtox commented:
Rebased, let's see what tests say

FabienPapet commented on src/Core/Util/Entity/EntityNameDuplicator.php:
```diff
@@ -0,0 +1,100 @@
+            $langId = (int) $langId;
+
+            // Add copy word to the string
+            $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```
`Language::getLocaleById($langId)` should be replaced y dependency injection

FabienPapet commented on src/Core/Util/Entity/EntityNameDuplicator.php:
```diff
@@ -0,0 +1,100 @@
+    public function getNewName(string $currentName, int $maxLength, int $langId): string
+    {
+        // Add copy word to the string
+        $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```
same here, use DI

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/util/entity.yml:
```diff
@@ -0,0 +1,9 @@
+services:
+  _defaults:
+    public: true
```
```suggestion
    public: false
```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/util/entity.yml:
```diff
@@ -0,0 +1,9 @@
+  _defaults:
+    public: true
+
+  prestashop.core.util.entity.entity_name_duplicator:
```
Update service definition to use FQCN

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/core/util/entity.yml:
```diff
@@ -0,0 +1,9 @@
+
+  prestashop.core.util.entity.entity_name_duplicator:
+    class: 'PrestaShop\PrestaShop\Core\Util\Entity\EntityNameDuplicator'
+    arguments:
```
```suggestion
    autowire: true
    arguments:
```

FabienPapet commented on src/PrestaShopBundle/Entity/Repository/CmsPageRepository.php:
```diff
@@ -0,0 +1,83 @@
+use PrestaShop\PrestaShop\Core\Repository\AbstractObjectModelRepository;
+use PrestaShop\PrestaShop\Core\Util\Entity\EntityNameDuplicator;
+
+class CmsPageRepository extends AbstractObjectModelRepository
```
This should have been inside an adapter

Hlavtox commented:
Ping @zuk3975 @FabienPapet @kpodemski for review, feedback should be addressed.

jolelievre commented on src/Core/Grid/Definition/Factory/CmsPageDefinitionFactory.php:
```diff
@@ -166,6 +168,16 @@ protected function getColumns()
+                        'route' => 'admin_cms_pages_view',
+                        'route_param_name' => 'cmsPageId',
+                        'route_param_field' => 'id_cms',
+                        'use_inline_display' => true,
```
Seems like this action opens a preview of the CMS page, then maybe it should be opened in another tab
```suggestion
                        'use_inline_display' => true,
                        'target' => '_blank',
```

jolelievre commented on src/PrestaShopBundle/Resources/config/services/adapter/product.yml:
```diff
@@ -405,6 +405,7 @@ services:
     arguments:
       $connection: '@doctrine.dbal.default_connection'
       $dbPrefix: '%database_prefix%'
+      $entityNameDuplicator: '@PrestaShop\PrestaShop\Adapter\Util\Entity\EntityNameDuplicator'
```
I think the autowire would have found it automatically since there's only one instance, but not blocking

jolelievre commented on src/PrestaShopBundle/Resources/config/services/core/util/entity.yml:
```diff
@@ -0,0 +1,11 @@
+    public: false
+
+  PrestaShop\PrestaShop\Adapter\Util\Entity\EntityNameDuplicator:
+    class: 'PrestaShop\PrestaShop\Adapter\Util\Entity\EntityNameDuplicator'
```
```suggestion
```
This is redundant with the FQCN notation

jolelievre reviewed (commented):
Thanks @Hlavtox 

a few non blocking comments. However CS fixer is not happy...

zuk3975 commented on src/Adapter/Util/Entity/EntityNameDuplicator.php:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @return void
+     */
+    private function buildLanguages(): void
```
noticed just now, minor stuff, can change it during other fixes.
```suggestion
    private function buildLocales(): void
```

Hlavtox committed:
Rename function

kpodemski committed:
Fix PHPStan

zuk3975 commented on src/PrestaShopBundle/Entity/Repository/CmsPageRepository.php:
```diff
@@ -26,7 +26,7 @@
 
 namespace PrestaShopBundle\Entity\Repository;
```
This is wrong namespace for object model.
If you want using object model, you must introduce the ObjectModel repository under adapter.
This namespace however is supposed to be used by doctrine entities only.
That is also why phpstan is complaining about using legacy CMS object model inside bundle namespace.

Hlavtox commented:
As nobody could help to with tests for half a year, I am closing this.
